### PR TITLE
Feature/555 taker flow

### PR DESCRIPTION
**Pull request by coltraneyan**
Would appreciate some eyes on this. 

Done
1. Have set up to pull FullOrder data from URL and pass down to component + children
2. Altered stringToSigDecimals function slightly to encompass an edge case

Issues:
1. SettingsPopover calls AppRouteParams, which only returns "order", not the full URL on language change. 

Questions:
1. Majority of component relies on tokenLists from a connected wallet. I'm thinking a loading spinner + "Please connect wallet" overlay to cover these cases?
2. does OrderStatus persist in store? Is there a way to know whether an order has been cancelled or taken?
3. what function will give me a correct txHash for etherscan url? 

Branched from 544/maker-flow. Final PR will rebase etc.

### CODE REVIEW

**Comment by piersss** [comment]
> Would appreciate some eyes on this.
> 
> Done
> 
> 1. Have set up to pull FullOrder data from URL and pass down to component + children
> 2. Altered stringToSigDecimals function slightly to encompass an edge case
> 
> Issues:
> 
> 1. SettingsPopover calls AppRouteParams, which only returns "order", not the full URL on language change.
> 
> Questions:
> 
> 1. Majority of component relies on tokenLists from a connected wallet. I'm thinking a loading spinner + "Please connect wallet" overlay to cover these cases?
> 2. does OrderStatus persist in store? Is there a way to know whether an order has been cancelled or taken?
> 3. what function will give me a correct txHash for etherscan url?
> 
> Branched from 544/maker-flow. Final PR will rebase etc.


> 1. SettingsPopover calls AppRouteParams, which only returns "order", not the full URL on language change.

`AppRouteParams` is only for the SwapWidget, should definitely rename that to avoid confusion.

You should just use `useParams` from `react-dom`. I will add changes to the _make_ branch to make it work. Then you should just use this to `OrderDetailWidget`:

`To const params = useParams<{ compressedOrder: 'string' }>();`

> 1. Majority of component relies on tokenLists from a connected wallet. I'm thinking a loading spinner + "Please connect wallet" overlay to cover these cases?

Refer to the Make and Swap widget how to do this.
 
> 2. does OrderStatus persist in store? Is there a way to know whether an order has been cancelled or taken?

Good question. Please ask Don or on the dev meeting tuesday.

> 3. what function will give me a correct txHash for etherscan url?

Same! Not sure how to know an order is taken and what the txHash is.

You can still prepare for the canceled / taken scenario's though. So just hardcode it for now for testing.

**Review by piersss — CHANGES_REQUESTED** [review]

File: src/components/OrderDetailWidget/OrderDetailWidget.tsx [hunk 1]
```diff
@@ -2,68 +2,115 @@ import React, { FC } from "react";
...
         onChangeTokenClick={() => {}}
         onMaxButtonClick={() => {}}
       />
-      <StyledInfoButtons
-        ownerIsCurrentUser
-        onFeeButtonClick={toggleShowFeeInfo}
-      />
+      {senderToken && signerToken && (
+        <StyledInfoButtons
+          ownerIsCurrentUser
```
[the remark this answers is not in the document]
Comment (piersss) [reply src/components/OrderDetailWidget/OrderDetailWidget.tsx:104]: But I like what you're doing with giving `InfoButtons` and `ActionButtons` boolean properties like `ownerIsCurrentUser` and `hasInsufficientBalance`. We should have a bunch of these eventually.

File: src/components/OrderDetailWidget/OrderDetailWidget.tsx [hunk 2]
```diff
@@ -1,73 +1,169 @@
...
-  const { account } = useWeb3React<Web3Provider>();
-
+  const { setShowWalletList } = useContext(InterfaceContext);
+  const { active, account, error: web3Error } = useWeb3React<Web3Provider>();
   const [showFeeInfo, toggleShowFeeInfo] = useToggle(false);
+  const senderToken = useTokenInfo(order.senderToken);
+  const signerToken = useTokenInfo(order.signerToken);
+  const hasInsufficientTokenBalance = useInsufficientBalance(
+    signerToken,
+    ethers.utils.formatUnits(order.signerAmount, signerToken?.decimals!)
```
Comment (piersss) [inline src/components/OrderDetailWidget/OrderDetailWidget.tsx:46]: `signerToken?.decimals!` what if signerToken is null? Things will break.

File: src/components/OrderDetailWidget/OrderDetailWidget.tsx [hunk 3]
```diff
@@ -1,73 +1,169 @@
...
 
-// Dummy data, this data will come from store later.
-const expiry = addDays(new Date(), 1);
-const orderStatus = OrderStatus.open;
-const orderType = OrderType.private;
-const recipientAddress = nativeCurrencyAddress;
-const transactionLink =
-  "https://etherscan.io/tx/0x53ddb90435ab46d96035da71b4ff8a26bb0a682a7f2327aaeb8ff3848406204f";
+interface OrderDetailWidgetProps {
+  status: "idle" | "not-found" | "open" | "taken" | "canceled";
```
Comment (piersss) [inline src/components/OrderDetailWidget/OrderDetailWidget.tsx:32]:
```suggestion
```
Get it from the store:

`const { status } = useAppSelector(selectTakeOtcReducer);`

File: src/components/OrderDetailWidget/OrderDetailWidget.tsx [hunk 4]
```diff
@@ -1,73 +1,169 @@
...
+  const { active, account, error: web3Error } = useWeb3React<Web3Provider>();
   const [showFeeInfo, toggleShowFeeInfo] = useToggle(false);
+  const senderToken = useTokenInfo(order.senderToken);
+  const signerToken = useTokenInfo(order.signerToken);
+  const hasInsufficientTokenBalance = useInsufficientBalance(
+    signerToken,
+    ethers.utils.formatUnits(order.signerAmount, signerToken?.decimals!)
+  );
+
+  const orderStatus = () => {
```
Comment (piersss) [inline src/components/OrderDetailWidget/OrderDetailWidget.tsx:49]:
```suggestion
  const getOrderStatus = () => {
```

Put in OrderDetailWidget/helpers/index.ts

**Review by piersss — CHANGES_REQUESTED** [review]
🕶️

File: src/components/OrderDetailWidget/OrderDetailWidget.tsx (hunk 2 above)
Comment (coltraneyan) [reply src/components/OrderDetailWidget/OrderDetailWidget.tsx:46]: checked out your implementation in MakeWidget. do you see the optimal solution as something like ```ethers.utils.formatUnits(order.signerAmount, signerToken?.decimals || "ether")``` or would it be better to handle in useTokenInfo, by always sending a ```defaultAddress```?

Comment (piersss) [reply src/components/OrderDetailWidget/OrderDetailWidget.tsx:46]: `decimals` should be a number, we could default it to 18 but then the user could potentially see a jumping number on the first frame.

And when `useTokenInfo` defaults to a `defaultAddress` you'll have the same problem. A token that's initially ether and then jumps to some other token. So the best way is to default to `undefined`:

```
const signerAmount = useMemo(() => {
    return signerToken ? ethers.utils.formatUnits(order.signerAmount, signerToken.decimals) : undefined
  }, [order, signerToken]);
  const hasInsufficientTokenBalance = useInsufficientBalance(
    signerToken,
    signerAmount,
  );
```

And then adjust hasInsufficientTokenBalance line 12: `amount?: string`.

And then to make things more clean I would add a `useFormattedTokenAmount` hook:

```
const useFormattedTokenAmount = (
  amount?: string,
  decimals?: number,
): string | undefined => {
  return useMemo(() => {
    if (!decimals || !amount) {
      return undefined;
    }

    return ethers.utils.formatUnits(amount, decimals);
  }, [amount, decimals]);
};

export default useFormattedTokenAmount;
```

And then OrderDetailWidget.tsx will look a lot cleaner.

```
  const signerAmount = useFormattedTokenAmount(order.signerAmount, signerToken?.decimals);
  const hasInsufficientTokenBalance = useInsufficientBalance(
    signerToken,
    signerAmount,
  );
```

You can do the same for `senderAmount` and then reduce some code for `const baseAmount` and `const quoteAmount`.

File: src/components/OrderDetailWidget/OrderDetailWidget.tsx (hunk 4 above)
Comment (piersss) [reply src/components/OrderDetailWidget/OrderDetailWidget.tsx]: _Put in OrderDetailWidget/helpers/index.ts_

Comment (coltraneyan) [reply src/components/OrderDetailWidget/OrderDetailWidget.tsx]: pushed

**Review by piersss — CHANGES_REQUESTED** [review]
Did some testing, taker scenario's seems to be working. Just some small stuff:

Screenshot 2022-09-22 at 20 29 38

- "New swap button" is disabled.
- Conversion rate is not correct.

Screenshot 2022-09-22 at 21 15 03

- "Close" button should be "Back"

**Comment by piersss** [comment]
And some warnings in console:

```
src/components/OrderDetailWidget/OrderDetailWidget.tsx
  Line 85:6:  React Hook useMemo has a missing dependency: 'signerAmount'. Either include it or remove the dependency array  react-hooks/exhaustive-deps
  Line 91:6:  React Hook useMemo has a missing dependency: 'senderAmount'. Either include it or remove the dependency array  react-hooks/exhaustive-deps
```

File: src/components/OrderDetailWidget/subcomponents/InfoButtons/InfoButtons.styles.tsx [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import styled from "styled-components/macro";
 
 import { LargePillButton } from "../../../../styled-components/Pill/Pill";
+import { ThemeButton } from "../../../SettingsPopover/SettingsPopover.styles";
```
Comment (piersss) [inline src/components/OrderDetailWidget/subcomponents/InfoButtons/InfoButtons.styles.tsx:4]:
```suggestion
```

**Review by piersss — CHANGES_REQUESTED** [review]

File: src/components/OrderDetailWidget/subcomponents/InfoButtons/InfoButtons.styles.tsx (hunk 5 above)
Comment (piersss) [reply src/components/OrderDetailWidget/subcomponents/InfoButtons/InfoButtons.styles.tsx:4]: 😬

**Review by piersss — APPROVED** [review]
You can squash and merge. 🎊  Please keep commit message to 1 line.

**Pull request merged**